File: Solutions/Python/symmetric_tree.py
"""
    [EASY]
    101. Symmetric Tree

    Concepts:
    - binary tree
    - breadth-first search

    Stats:
        Runtime | 0 ms      [Beats 100%]
        Memory  | 16.55 MB  [Beats 100%]
"""

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isSymmetric(self, root: Optional[TreeNode]) -> bool:
        """
            Search both subtrees using BFS
            Compare each level to see if reverse of each other
        """

        def bfs(root):
            queue = deque([root])
            levels = []

            while queue:
                n = len(queue)
                level = []
                for _ in range(n):
                    curr = queue.popleft()
                    if curr:
                        level.append(curr.val)
                        queue.append(curr.left)
                        queue.append(curr.right)
                    else:
                        level.append(None)

                levels.append(level)

            return levels

        for level in bfs(root):
            if level != level[::-1]:
                return False

        return True

File: Solutions/Cpp/symmetric_tree.cpp
/*
    [EASY]
    101. Symmetric Tree

    Concepts:
    - binary tree
    - depth-first search

    Stats:
        Runtime | 0 ms      [Beats 100%]
        Memory  | 19.04 MB  [Beats 5.90%]
*/

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool isSymmetric(TreeNode* root) {
        std::function<bool(TreeNode*, TreeNode*)> dfs;

        dfs = [&](TreeNode* node1, TreeNode* node2){
            if (node1 == nullptr && node2 == nullptr)
                return true;
            if (node1 == nullptr || node2 == nullptr)
                return false;

            return (
                node1->val == node2->val &&
                dfs(node1->left, node2->right) &&
                dfs(node1->right, node2->left)
            );
        };

        return dfs(root->left, root->right);
    }
};